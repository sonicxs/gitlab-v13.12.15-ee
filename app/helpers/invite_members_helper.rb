# frozen_string_literal: true

module InviteMembersHelper
  include Gitlab::Utils::StrongMemoize

  def can_invite_members_for_project?(project)
    Feature.enabled?(:invite_members_group_modal, project.group) && can_manage_project_members?(project)
  end

  def can_invite_group_for_project?(project)
    Feature.enabled?(:invite_members_group_modal, project.group) && project.allowed_to_share_with_group?
  end

  def directly_invite_members?
    strong_memoize(:directly_invite_members) do
      can_import_members?
    end
  end

  def invite_group_members?(group)
    experiment_enabled?(:invite_members_empty_group_version_a) && Ability.allowed?(current_user, :admin_group_member, group)
  end

  def dropdown_invite_members_link(form_model)
    link_to invite_members_url(form_model),
            data: {
              'track-event': 'click_link',
              'track-label': tracking_label,
              'track-property': experiment_tracking_category_and_group(:invite_members_new_dropdown)
            } do
      invite_member_link_content
    end
  end

  def invite_accepted_notice(member)
    case member.source
    when Project
      _("You have been granted %{member_human_access} access to project %{name}.") %
        { member_human_access: member.human_access, name: member.source.name }
    when Group
      _("You have been granted %{member_human_access} access to group %{name}.") %
        { member_human_access: member.human_access, name: member.source.name }
    end
  end

  private

  def invite_members_url(form_model)
    case form_model
    when Project
      project_project_members_path(form_model)
    when Group
      group_group_members_path(form_model)
    end
  end

  def invite_member_link_content
    text = s_('InviteMember|Invite members')

    return text unless experiment_enabled?(:invite_members_new_dropdown)

    "#{text} #{emoji_icon('shaking_hands', 'aria-hidden': true, class: 'gl-font-base gl-vertical-align-baseline')}".html_safe
  end
end
