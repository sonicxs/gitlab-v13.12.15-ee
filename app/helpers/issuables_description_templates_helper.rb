# frozen_string_literal: true

module IssuablesDescriptionTemplatesHelper
  include Gitlab::Utils::StrongMemoize
  include GitlabRoutingHelper

  def template_dropdown_tag(issuable, &block)
    selected_template = selected_template(issuable)
    title = selected_template || "Choose a template"
    options = {
      toggle_class: 'js-issuable-selector',
      title: title,
      filter: true,
      placeholder: 'Filter',
      footer_content: true,
      data: {
        data: issuable_templates(ref_project, issuable.to_ability_name),
        field_name: 'issuable_template',
        selected: selected_template,
        project_id: ref_project.id
      }
    }

    dropdown_tag(title, options: options) do
      capture(&block)
    end
  end

  def issuable_templates(project, issuable_type)
    @template_types ||= {}
    @template_types[project.id] ||= {}
    @template_types[project.id][issuable_type] ||= TemplateFinder.all_template_names_hash_or_array(project, issuable_type)
  end

  def issuable_templates_names(issuable)
    all_templates = issuable_templates(ref_project, issuable.to_ability_name)

    if ref_project.inherited_issuable_templates_enabled?
      all_templates.values.flatten.map { |tpl| tpl[:name] if tpl[:project_id] == ref_project.id }.compact.uniq
    else
      all_templates.map { |template| template[:name] }
    end
  end

  def selected_template(issuable)
    params[:issuable_template] if issuable_templates_names(issuable).any? { |tmpl_name| tmpl_name == params[:issuable_template] }
  end

  def template_names_path(parent, issuable)
    return '' unless parent.is_a?(Project)

    project_template_names_path(parent, template_type: issuable.to_ability_name)
  end
end
