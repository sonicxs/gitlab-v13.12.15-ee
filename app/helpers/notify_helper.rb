# frozen_string_literal: true

module NotifyHelper
  def merge_request_reference_link(entity, *args)
    link_to(entity.to_reference, merge_request_url(entity, *args))
  end

  def issue_reference_link(entity, *args, full: false)
    link_to(entity.to_reference(full: full), issue_url(entity, *args))
  end

  def invited_role_description(role_name)
    case role_name
    when "Guest"
      s_("InviteEmail|As a guest, you can view projects, leave comments, and create issues.")
    when "Reporter"
      s_("InviteEmail|As a reporter, you can view projects and reports, and leave comments on issues.")
    when "Developer"
      s_("InviteEmail|As a developer, you have full access to projects, so you can take an idea from concept to production.")
    when "Maintainer"
      s_("InviteEmail|As a maintainer, you have full access to projects. You can push commits to the default branch and deploy to production.")
    when "Owner"
      s_("InviteEmail|As an owner, you have full access to projects and can manage access to the group, including inviting new members.")
    when "Minimal Access"
      s_("InviteEmail|As a user with minimal access, you can view the high-level group from the UI and API.")
    end
  end

  def invited_to_description(source)
    case source
    when "project"
      s_('InviteEmail|Projects can be used to host your code, track issues, collaborate on code, and continuously build, test, and deploy your app with built-in GitLab CI/CD.')
    when "group"
      s_('InviteEmail|Groups assemble related projects together and grant members access to several projects at once.')
    end
  end
end
