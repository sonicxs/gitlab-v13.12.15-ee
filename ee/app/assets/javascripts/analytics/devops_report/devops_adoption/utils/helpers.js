import { isToday } from '~/lib/utils/datetime_utility';

/**
 * A helper function which accepts the segments,
 *
 * @param {Object} params the segment data, timestamp and check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ segments, timestamp, openModal }) => {
  if (openModal) {
    return false;
  } else if (!segments.length) {
    return true;
  }

  const anyPendingSegments = segments.some((node) => node.latestSnapshot === null);
  const dataNotRefreshedToday = !isToday(new Date(timestamp));

  return anyPendingSegments || dataNotRefreshedToday;
};
