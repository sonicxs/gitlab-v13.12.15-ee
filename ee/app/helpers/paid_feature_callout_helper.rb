# frozen_string_literal: true

# NOTE: This is largely mimicking the structure created as part of the
# TrialStatusWidgetHelper (ee/app/helpers/trial_status_widget_helper.rb), & it
# is utilizing a few methods (including private ones) from that helper as well.
module PaidFeatureCalloutHelper
  def run_highlight_paid_features_during_active_trial_experiment(group, &block)
    experiment(:highlight_paid_features_during_active_trial, group: group) do |e|
      e.exclude! unless billing_plans_and_trials_available?
      e.exclude! unless group && eligible_for_trial_upgrade_callout?(group)
      e.use { nil } # control gets nothing new added to the existing UI
      e.try(&block)
    end
  end

  def paid_feature_badge_data_attrs(feature_name)
    base_attrs = base_paid_feature_data_attrs(feature_name)

    base_attrs.merge({
      id: feature_callout_container_id(feature_name)
    })
  end

  def paid_feature_popover_data_attrs(group:, feature_name:)
    base_attrs = base_paid_feature_data_attrs(feature_name)
    container_id = feature_callout_container_id(feature_name)

    base_attrs.merge({
      container_id: container_id,
      days_remaining: group.trial_days_remaining,
      href_compare_plans: group_billings_path(group),
      href_upgrade_to_paid: premium_subscription_path_for_group(group),
      plan_name_for_trial: group.gitlab_subscription&.plan_title,
      plan_name_for_upgrade: 'Premium',
      target_id: container_id
    })
  end

  private

  def feature_callout_container_id(feature_name)
    "#{feature_name.parameterize}-callout"
  end

  def base_paid_feature_data_attrs(feature_name)
    { feature_name: feature_name }
  end

  def premium_subscription_path_for_group(group)
    # NOTE: We are okay hard-coding the production value for the Premium 1-year
    # SaaS plan ID while this is all part of an active experiment. If & when the
    # experiment is deemed a success, part of the clean-up effort will be to
    # pull the value directly from the CustomersDot API. Value taken from
    # https://gitlab.com/gitlab-org/customers-gitlab-com/blob/7177f13c478ef623b779d6635c4a58ee650b7884/config/application.yml#L186
    # Cleanup issue: https://gitlab.com/gitlab-org/gitlab/-/issues/330987
    zuora_premium_plan_id = '2c92a00d76f0d5060176f2fb0a5029ff'

    new_subscriptions_path(namespace_id: group.id, plan_id: zuora_premium_plan_id)
  end
end
