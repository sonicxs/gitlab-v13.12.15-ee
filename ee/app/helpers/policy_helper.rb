# frozen_string_literal: true

module PolicyHelper
  def policy_details(project, policy = nil, environment = nil)
    return unless project

    details = details(project)
    return details unless policy && environment

    edit_details = {
      policy: policy.to_json,
      environment_id: environment.id
    }
    details.merge(edit_details)
  end

  def threat_monitoring_alert_details_data(project, alert_iid)
    {
      'alert-id' => alert_iid,
      'project-path' => project.full_path,
      'project-id' => project.id,
      'project-issues-path' => project_issues_path(project),
      'page' => 'THREAT_MONITORING'
    }
  end

  private

  def details(project)
    {
      network_policies_endpoint: project_security_network_policies_path(project),
      configure_agent_help_path: help_page_url('user/clusters/agent/repository.html'),
      create_agent_help_path: help_page_url('user/clusters/agent/index.md', anchor: 'create-an-agent-record-in-gitlab'),
      environments_endpoint: project_environments_path(project),
      project_path: project.full_path,
      project_id: project.id,
      threat_monitoring_path: project_threat_monitoring_path(project)
    }
  end
end
