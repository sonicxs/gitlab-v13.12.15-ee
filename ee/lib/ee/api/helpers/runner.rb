# frozen_string_literal: true

module EE
  module API
    module Helpers
      module Runner
        extend ::Gitlab::Utils::Override

        override :current_job
        def current_job
          id = params[:id]

          if id
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :build, id)
          end

          super
        end

        override :current_runner
        def current_runner
          token = params[:token]

          if token
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :runner, token)
          end

          super
        end

        override :track_ci_minutes_usage!
        def track_ci_minutes_usage!(build, runner)
          ::Ci::Minutes::TrackLiveConsumptionService.new(build).execute
        end
      end
    end
  end
end
