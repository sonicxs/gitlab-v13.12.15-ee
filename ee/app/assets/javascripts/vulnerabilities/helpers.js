import { isAbsolute, isSafeURL } from '~/lib/utils/url_utility';
import { REGEXES, gidPrefix, uidPrefix } from './constants';

// Get the issue in the format expected by the descendant components of related_issues_block.vue.
export const getFormattedIssue = (issue) => ({
  ...issue,
  reference: `#${issue.iid}`,
  path: issue.web_url,
});

export const getAddRelatedIssueRequestParams = (reference, defaultProjectId) => {
  let issueId = reference;
  let projectId = defaultProjectId;

  // If the reference is an issue number, parse out just the issue number.
  if (REGEXES.ISSUE_FORMAT.test(reference)) {
    [, issueId] = REGEXES.ISSUE_FORMAT.exec(reference);
  }
  // If the reference is an absolute URL and matches the issues URL format, parse out the project and issue.
  else if (isSafeURL(reference) && isAbsolute(reference)) {
    const { pathname } = new URL(reference);

    if (REGEXES.LINK_FORMAT.test(pathname)) {
      [, projectId, issueId] = REGEXES.LINK_FORMAT.exec(pathname);
    }
  }

  return { target_issue_iid: issueId, target_project_id: projectId };
};

export const normalizeGraphQLVulnerability = (vulnerability) => {
  if (!vulnerability) {
    return null;
  }

  const newVulnerability = { ...vulnerability };

  if (vulnerability.id) {
    newVulnerability.id = vulnerability.id.replace(gidPrefix, '');
  }

  if (vulnerability.state) {
    newVulnerability.state = vulnerability.state.toLowerCase();
  }

  ['confirmed', 'resolved', 'dismissed'].forEach((state) => {
    if (vulnerability[`${state}By`]?.id) {
      newVulnerability[`${state}ById`] = vulnerability[`${state}By`].id.replace(uidPrefix, '');
      delete newVulnerability[`${state}By`];
    }
  });

  return newVulnerability;
};
