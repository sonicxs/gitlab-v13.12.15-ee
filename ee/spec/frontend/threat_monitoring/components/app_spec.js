import { shallowMount } from '@vue/test-utils';
import ThreatMonitoringAlerts from 'ee/threat_monitoring/components/alerts/alerts.vue';
import ThreatMonitoringApp from 'ee/threat_monitoring/components/app.vue';
import NetworkPolicyList from 'ee/threat_monitoring/components/network_policy_list.vue';
import NoEnvironmentEmptyState from 'ee/threat_monitoring/components/no_environment_empty_state.vue';
import ThreatMonitoringFilters from 'ee/threat_monitoring/components/threat_monitoring_filters.vue';
import createStore from 'ee/threat_monitoring/store';
import { TEST_HOST } from 'helpers/test_constants';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';

const defaultEnvironmentId = 3;
const documentationPath = '/docs';
const newPolicyPath = '/policy/new';
const emptyStateSvgPath = '/svgs';
const wafNoDataSvgPath = '/waf-no-data-svg';
const networkPolicyNoDataSvgPath = '/network-policy-no-data-svg';
const environmentsEndpoint = `${TEST_HOST}/environments`;
const wafStatisticsEndpoint = `${TEST_HOST}/waf`;
const networkPolicyStatisticsEndpoint = `${TEST_HOST}/network_policy`;

describe('ThreatMonitoringApp component', () => {
  let store;
  let wrapper;

  const factory = ({ propsData, provide = {}, state, stubs = {} } = {}) => {
    store = createStore();
    Object.assign(store.state.threatMonitoring, {
      environmentsEndpoint,
      wafStatisticsEndpoint,
      networkPolicyStatisticsEndpoint,
      ...state,
    });

    jest.spyOn(store, 'dispatch').mockImplementation();

    wrapper = extendedWrapper(
      shallowMount(ThreatMonitoringApp, {
        propsData: {
          defaultEnvironmentId,
          emptyStateSvgPath,
          wafNoDataSvgPath,
          networkPolicyNoDataSvgPath,
          newPolicyPath,
          ...propsData,
        },
        provide: {
          documentationPath,
          glFeatures: { threatMonitoringAlerts: false },
          ...provide,
        },
        store,
        stubs,
      }),
    );
  };

  const findAlertsView = () => wrapper.find(ThreatMonitoringAlerts);
  const findNetworkPolicyList = () => wrapper.find(NetworkPolicyList);
  const findFilters = () => wrapper.find(ThreatMonitoringFilters);
  const findWafSection = () => wrapper.find({ ref: 'wafSection' });
  const findNetworkPolicySection = () => wrapper.find({ ref: 'networkPolicySection' });
  const findNoEnvironmentEmptyStates = () => wrapper.findAll(NoEnvironmentEmptyState);
  const findNetworkPolicyTab = () => wrapper.find({ ref: 'networkPolicyTab' });
  const findAlertTab = () => wrapper.findByTestId('threat-monitoring-alerts-tab');
  const findStatisticsTab = () => wrapper.findByTestId('threat-monitoring-statistics-tab');

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe.each([-1, NaN, Math.PI])(
    'given an invalid default environment id of %p',
    (invalidEnvironmentId) => {
      beforeEach(() => {
        factory({
          propsData: {
            defaultEnvironmentId: invalidEnvironmentId,
          },
          stubs: { GlTabs: false },
        });
      });

      it('dispatches no actions', () => {
        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('shows the "no environment" empty state', () => {
        expect(findNoEnvironmentEmptyStates().length).toBe(2);
      });

      it('shows the tabs', () => {
        expect(findNetworkPolicyTab().exists()).toBe(true);
        expect(findStatisticsTab().exists()).toBe(true);
      });

      it('does not show the network policy list', () => {
        expect(findNetworkPolicyList().exists()).toBe(false);
      });

      it('does not show the threat monitoring section', () => {
        expect(findNetworkPolicySection().exists()).toBe(false);
      });
    },
  );

  describe('given there is a default environment with data', () => {
    beforeEach(() => {
      factory();
    });

    it('dispatches the setCurrentEnvironmentId and fetchEnvironments actions', () => {
      expect(store.dispatch.mock.calls).toEqual([
        ['threatMonitoring/setCurrentEnvironmentId', defaultEnvironmentId],
        ['threatMonitoring/fetchEnvironments', undefined],
      ]);
    });

    it('shows the filter bar', () => {
      expect(findFilters().exists()).toBe(true);
    });

    it('renders the waf section', () => {
      expect(findWafSection().element).toMatchSnapshot();
    });

    it('renders the network policy section', () => {
      expect(findNetworkPolicySection().element).toMatchSnapshot();
    });

    it('renders the network policy tab', () => {
      expect(findNetworkPolicyTab().element).toMatchSnapshot();
    });

    it('does not show the alert tab', () => {
      expect(findAlertTab().exists()).toBe(false);
    });
  });

  describe('alerts tab', () => {
    beforeEach(() => {
      factory({ provide: { glFeatures: { threatMonitoringAlerts: true } } });
    });
    it('shows the alerts tab', () => {
      expect(findAlertTab().exists()).toBe(true);
    });
    it('shows the default alerts component', () => {
      expect(findAlertsView().exists()).toBe(true);
    });
  });
});
