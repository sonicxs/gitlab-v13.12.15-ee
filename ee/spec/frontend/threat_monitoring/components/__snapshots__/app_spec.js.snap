// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreatMonitoringApp component given there is a default environment with data renders the network policy section 1`] = `
<threat-monitoring-section-stub
  anomaloustitle="Dropped Packets"
  chartemptystatesvgpath="/network-policy-no-data-svg"
  chartemptystatetext="Container Network Policies are not installed or have been disabled. To view this data, ensure your Network Policies are installed and enabled for your cluster."
  chartemptystatetitle="Container NetworkPolicies not detected"
  documentationanchor="container-network-policy"
  documentationpath="/docs"
  nominaltitle="Total Packets"
  storenamespace="threatMonitoringNetworkPolicy"
  subtitle="Packet Activity"
  title="Container Network Policy"
  ylegend="Operations Per Second"
/>
`;

exports[`ThreatMonitoringApp component given there is a default environment with data renders the network policy tab 1`] = `
<gl-tab-stub
  title="Policies"
  titlelinkclass=""
>
  <network-policy-list-stub
    documentationpath="/docs"
    newpolicypath="/policy/new"
  />
</gl-tab-stub>
`;

exports[`ThreatMonitoringApp component given there is a default environment with data renders the waf section 1`] = `
<threat-monitoring-section-stub
  anomaloustitle="Anomalous Requests"
  chartemptystatesvgpath="/waf-no-data-svg"
  chartemptystatetext="The firewall is not installed or has been disabled. To view this data, ensure the web application firewall is installed and enabled for your cluster."
  chartemptystatetitle="Application firewall not detected"
  documentationanchor="web-application-firewall"
  documentationpath="/docs"
  nominaltitle="Total Requests"
  storenamespace="threatMonitoringWaf"
  subtitle="Requests"
  title="Web Application Firewall"
  ylegend="Requests"
/>
`;
